rio/src/RIO/Prelude/RIO.hs

```diff
@@ -35,3 +50,104 @@ instance MonadUnliftIO (RIO env) where
 instance PrimMonad (RIO env) where
     type PrimState (RIO env) = PrimState IO
     primitive = RIO . ReaderT . const . primitive
+
+-- | Abstraction over how to read from and write to a reference
+--
+-- @since 0.1.3.0
+data SomeRef a
+  = SomeRef !(IO a) !(a -> IO ())
+
+-- | read from a SomeRef
+--
+-- @since 0.1.3.0
+readSomeRef :: SomeRef a -> IO a
+readSomeRef (SomeRef x _) = x
+
+-- | write to a SomeRef
+--
+-- @since 0.1.3.0
+writeSomeRef :: SomeRef a -> a -> IO ()
+writeSomeRef (SomeRef _ x) = x
+
+-- | modify a SomeRef
+--
+-- @since 0.1.3.0
+modifySomeRef :: SomeRef a -> (a -> a) -> IO ()
+modifySomeRef (SomeRef read write) f =
+  (f <$> read) >>= write
+
+ioRefToSomeRef :: IORef a -> SomeRef a
+ioRefToSomeRef ref = do
+  SomeRef (readIORef ref)
+          (\val -> modifyIORef' ref (\_ -> val))
+
+uRefToSomeRef :: Unbox a => URef RealWorld a -> SomeRef a
+uRefToSomeRef ref = do
+  SomeRef (readURef ref) (writeURef ref)
+
+-- | Environment values with stateful capabilities to SomeRef
+--
+-- @since 0.1.3.0
+class HasStateRef s env | env -> s where
+  stateRefL :: Lens' env (SomeRef s)
+
+-- | Identity state reference where the SomeRef is the env
+--
+-- @since 0.1.3.0
+instance HasStateRef a (SomeRef a) where
+  stateRefL = lens id (\_ x -> x)
+
+-- | Environment values with writing capabilities to SomeRef
+--
+-- @since 0.1.3.0
+class HasWriteRef w env | env -> w where
+  writeRefL :: Lens' env (SomeRef w)
+
+-- | Identity write reference where the SomeRef is the env
+--
+-- @since 0.1.3.0
+instance HasWriteRef a (SomeRef a) where
```

Comment:
@mauropalsgraaf  wondering what is the reason we have a different typeclass for the same purpose?

Reply:
If you would only use one typeclass, the somerefs have to always be the same, while now I can have an environment with two SomeRefs, one for writing and one for state. We could add a default instance something like `instance HasStateRef w env => HasWriteRef w env where` to have both on the same place right?